src/_pytest/unraisableexception.py

```diff
@@ -1,100 +1,131 @@
-                else:
-                    err_msg = "Exception ignored in"
-                msg = f"{err_msg}: {cm.unraisable.object!r}\n\n"
-                msg += "".join(
-                    traceback.format_exception(
-                        cm.unraisable.exc_type,
-                        cm.unraisable.exc_value,
-                        cm.unraisable.exc_traceback,
-                    )
-                )
-                warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
+    pass
+
+if sys.version_info < (3, 11):
+    from exceptiongroup import ExceptionGroup
+
 
+class UnraisableMeta(NamedTuple):
+    msg: str
+    cause_msg: str
+    exc_value: BaseException | None
 
-@pytest.hookimpl(wrapper=True, tryfirst=True)
-def pytest_runtest_setup() -> Generator[None]:
-    yield from unraisable_exception_runtest_hook()
 
+_unraisable_exceptions: collections.deque[UnraisableMeta | BaseException] = (
+    collections.deque()
+)
 
-@pytest.hookimpl(wrapper=True, tryfirst=True)
-def pytest_runtest_call() -> Generator[None]:
-    yield from unraisable_exception_runtest_hook()
 
+def collect_unraisable() -> None:
+    errors: list[pytest.PytestUnraisableExceptionWarning | RuntimeError] = []
+    meta = None
+    hook_error = None
+    try:
+        while True:
+            try:
+                meta = _unraisable_exceptions.pop()
+            except IndexError:
+                break
 
-@pytest.hookimpl(wrapper=True, tryfirst=True)
-def pytest_runtest_teardown() -> Generator[None]:
-    yield from unraisable_exception_runtest_hook()
+            if isinstance(meta, BaseException):
+                hook_error = RuntimeError("Failed to process unraisable exception")
+                hook_error.__cause__ = meta
+                errors.append(hook_error)
+                continue
+
+            msg = meta.msg
+            try:
+                warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
+            except pytest.PytestUnraisableExceptionWarning as e:
+                if meta.exc_value is not None:
+                    # exceptions have a better way to show the traceback, but
+                    # warnings do not, so hide the traceback from the msg and
+                    # set the cause so the traceback shows up in the right place
+                    e.args = (meta.cause_msg,)
+                    e.__cause__ = meta.exc_value
+                errors.append(e)
+
+        if len(errors) == 1:
+            raise errors[0]
+        if errors:
+            raise ExceptionGroup("multiple unraisable exception warnings", errors)
+    finally:
+        del errors, meta, hook_error
+
+
+def _cleanup(prev_hook: Callable[[sys.UnraisableHookArgs], object]) -> None:
+    try:
+        for i in range(5):
+            gc.collect()
+        collect_unraisable()
+    finally:
+        sys.unraisablehook = prev_hook
+
+
+def unraisable_hook(unraisable: sys.UnraisableHookArgs) -> None:
+    try:
```

Comment:
If I understand correctly according to the tests, the thing that might raise in this try block is the `format_exception`, is this right or is there somthing else?

Can you reduce the scope of the `try` to the minimal parts that can raise? This makes it more clear what's going on, and also avoids "over catching" in case there is some unrelated bug.